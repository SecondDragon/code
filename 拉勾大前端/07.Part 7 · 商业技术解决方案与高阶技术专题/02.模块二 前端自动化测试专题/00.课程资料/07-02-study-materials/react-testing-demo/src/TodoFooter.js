// import classNames from 'classnames'
import { useMemo } from 'react'
import { NavLink } from 'react-router-dom'

export default function TodoFooter ({ todos, removeAllDone, filterText }) {
  const remainingCount = useMemo(
    () => {
      return todos.filter((t) => !t.done).length
    },
    [ todos ]
  )

  return (
    <footer className="footer">
      {/* This should be `0 items left` by default */}
      <span className="todo-count">
        <strong data-testid="remaining">{remainingCount}</strong> item left
      </span>
      {/* Remove this if you don't implement routing */}
      <ul className="filters">
        <li>
          <NavLink activeClassName="selected" exact to="/" data-testid="link-all">
            All
          </NavLink>
        </li>
        <li>
          <NavLink activeClassName="selected" to="/active" data-testid="link-active">
            Active
          </NavLink>
        </li>
        <li>
          <NavLink activeClassName="selected" to="/completed" data-testid="link-completed">
            Completed
          </NavLink>
        </li>
      </ul>
      {/* Hidden if no completed items are left â†“ */}
      {todos.some(t => t.done) ? (
        <button className="clear-completed" data-testid="clear-completed" onClick={removeAllDone}>
          Clear completed
        </button>
      ) : null}
    </footer>
  )
}
