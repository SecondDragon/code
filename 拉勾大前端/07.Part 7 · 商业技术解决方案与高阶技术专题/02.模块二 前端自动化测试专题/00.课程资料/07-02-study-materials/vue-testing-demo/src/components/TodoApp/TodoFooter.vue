<template>
  <footer class="footer">
    <!-- This should be `0 items left` by default -->
    <span class="todo-count"
      ><strong data-testid="done-todos-count">{{ doneTodosCount }}</strong> item
      left</span
    >
    <!-- Remove this if you don't implement routing -->
    <ul class="filters">
      <li>
        <!-- <a class="selected" href="#/">All</a> -->
        <router-link data-testid="link-all" exact to="/">All</router-link>
      </li>
      <li>
        <!-- <a href="#/active">Active</a> -->
        <router-link data-testid="link-active" to="/active">Active</router-link>
      </li>
      <li>
        <!-- <a href="#/completed">Completed</a> -->
        <router-link data-testid="link-completed" to="/completed">Completed</router-link>
      </li>
    </ul>
    <!-- Hidden if no completed items are left â†“ -->
    <button
      v-if="isClearCompletedShow"
      class="clear-completed"
      data-testid="clear-completed"
      @click="$emit('clear-completed')"
    >Clear completed</button>
  </footer>
</template>

<script>
export default {
  name: 'TodoFooter',
  props: {
    todos: {
      type: Array,
      required: true
    }
  },
  computed: {
    doneTodosCount () {
      return this.todos.filter(t => !t.done).length
    },
    isClearCompletedShow () {
      return this.todos.find(t => t.done)
    }
  }
}
</script>
