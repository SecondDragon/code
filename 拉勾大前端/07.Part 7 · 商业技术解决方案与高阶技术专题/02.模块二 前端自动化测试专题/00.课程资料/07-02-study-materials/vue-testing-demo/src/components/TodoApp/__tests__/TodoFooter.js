import { createLocalVue, mount } from '@vue/test-utils'
import TodoFooter from '@/components/TodoApp/TodoFooter.vue'
import VueRouter from 'vue-router'

const localVue = createLocalVue()
localVue.use(VueRouter)
const router = new VueRouter({
  linkActiveClass: 'selected'
})

describe('TodoFooter.vue', () => {
  /** @type {import('@vue/test-utils').Wrapper} */
  let wrapper = null

  beforeEach(async () => {
    const todos = [
      { id: 1, text: 'eat', done: false },
      { id: 2, text: 'play', done: true },
      { id: 3, text: 'sleep', done: false }
    ]
    wrapper = mount(TodoFooter, {
      propsData: {
        todos
      },
      localVue,
      router
    })
  })

  test('Done Todos Count', async () => {
    const count = wrapper.vm.todos.filter(t => !t.done).length
    const countEl = wrapper.find('[data-testid="done-todos-count"]')
    expect(Number.parseInt(countEl.text())).toBe(count)
  })

  test('Clear Completed Show', async () => {
    const button = wrapper.find('[data-testid="clear-completed"]')
    expect(button.exists()).toBeTruthy()

    // 清除所有任务的完成状态，判断 button 不存在
    wrapper = mount(TodoFooter, {
      propsData: {
        todos: [
          { id: 1, text: 'eat', done: false },
          { id: 2, text: 'play', done: false },
          { id: 3, text: 'sleep', done: false }
        ]
      },
      localVue,
      router
    })
    expect(wrapper.find('[data-testid="clear-completed"]').exists()).toBeFalsy()
  })

  test('Clear Completed', async () => {
    const button = wrapper.find('[data-testid="clear-completed"]')
    await button.trigger('click')
    expect(wrapper.emitted()['clear-completed']).toBeTruthy()
  })

  test('Router Link ActiveClass', async () => {
    const links = wrapper.findAllComponents({ name: 'RouterLink' })
    router.push('/active')
    await localVue.nextTick()
    for (let i = 0; i < links.length; i++) {
      const link = links.at(i)
      if (link.vm.to === '/active') {
        expect(link.classes()).toContain('selected')
      } else {
        expect(link.classes('selected')).toBeFalsy()
      }
    }

    router.push('/completed')
    await localVue.nextTick()
    for (let i = 0; i < links.length; i++) {
      const link = links.at(i)
      if (link.vm.to === '/completed') {
        expect(link.classes()).toContain('selected')
      } else {
        expect(link.classes('selected')).toBeFalsy()
      }
    }

    router.push('/')
    await localVue.nextTick()
    for (let i = 0; i < links.length; i++) {
      const link = links.at(i)
      if (link.vm.to === '/') {
        expect(link.classes()).toContain('selected')
      } else {
        expect(link.classes('selected')).toBeFalsy()
      }
    }
  })

  test('snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
})
