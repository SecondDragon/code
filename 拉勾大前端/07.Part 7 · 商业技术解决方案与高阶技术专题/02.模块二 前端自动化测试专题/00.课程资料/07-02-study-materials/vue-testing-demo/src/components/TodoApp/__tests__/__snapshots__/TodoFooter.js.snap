// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoFooter.vue snapshot 1`] = `
<footer class="footer"><span class="todo-count"><strong data-testid="done-todos-count">2</strong> item
    left</span>
  <ul class="filters">
    <li><a href="#/" aria-current="page" class="router-link-exact-active selected">All</a></li>
    <li><a href="#/active" class="">Active</a></li>
    <li><a href="#/completed" class="">Completed</a></li>
  </ul> <button data-testid="clear-completed" class="clear-completed">Clear completed</button>
</footer>
`;
